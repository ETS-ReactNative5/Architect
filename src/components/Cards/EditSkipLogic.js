import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators, compose } from 'redux';
import PropTypes from 'prop-types';
import { get, find } from 'lodash';
import { Button } from '../../ui/components';
import { actionCreators as stageActions } from '../../ducks/modules/protocol/stages';
import Card from './ProtocolCard';
import { getProtocol, getVariableRegistry } from '../../selectors/protocol';
import { Draft } from '../../behaviours';
import SkipLogicEditor from '../SkipLogicEditor';

const defaultLogic = {
  action: 'SKIP',
  operator: '',
  value: '',
  filter: {
    join: '',
    rules: [],
  },
};

class EditSkipLogic extends PureComponent {
  static propTypes = {
    show: PropTypes.bool,
    hasChanges: PropTypes.bool,
    stageId: PropTypes.string,
    onComplete: PropTypes.func,
    draft: PropTypes.any.isRequired,
    updateStage: PropTypes.func.isRequired,
    updateDraft: PropTypes.func.isRequired,
    variableRegistry: PropTypes.object.isRequired,
  };

  static defaultProps = {
    show: false,
    stageId: null,
    hasChanges: false,
    onComplete: () => {},
  }

  get buttons() {
    return this.props.hasChanges ?
      [<Button key="save" onClick={this.handleSave}>Continue</Button>] :
      [];
  }

  handleCancel = () => {
    this.props.onComplete();
  }

  handleSave = () => {
    const stageId = this.props.stageId;

    this.props.updateStage(
      stageId,
      {
        skipLogic: this.props.draft,
      },
    );

    this.props.onComplete();
  };

  handleChange = this.props.updateDraft;

  render() {
    const {
      show,
      draft,
      variableRegistry,
    } = this.props;

    return (
      <Card
        type="logic"
        buttons={this.buttons}
        show={show}
        onCancel={this.handleCancel}
      >
        <SkipLogicEditor
          onChange={this.handleChange}
          rules={draft}
          variableRegistry={variableRegistry}
        />
      </Card>
    );
  }
}

const mapStateToProps = (state, props) => {
  const stageId = get(props, 'match.params.id');
  const protocol = getProtocol(state);
  const stage = find(protocol.stages, ['id', stageId]);
  const logic = get(stage, 'skipLogic', defaultLogic);
  const variableRegistry = getVariableRegistry(state);

  return {
    stageId,
    draft: logic,
    variableRegistry,
  };
};

const mapDispatchToProps = dispatch => ({
  updateStage: bindActionCreators(stageActions.updateStage, dispatch),
});

export { EditSkipLogic };

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  Draft,
)(EditSkipLogic);
