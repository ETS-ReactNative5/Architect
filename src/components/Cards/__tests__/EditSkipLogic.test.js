/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { EditSkipLogic } from '../EditSkipLogic';

const mockProps = {
  draft: {},
  updateStage: () => {},
  updateDraft: () => {},
  rules: {},
  onChange: () => {},
  variableRegistry: {},
};

describe('<EditForm />', () => {
  it('can render', () => {
    const component = shallow(<EditSkipLogic {...mockProps} />);

    expect(component).toMatchSnapshot();
  });
});
