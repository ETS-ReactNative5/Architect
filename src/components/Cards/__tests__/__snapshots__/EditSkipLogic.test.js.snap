// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditForm /> can render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditSkipLogic
    draft={Object {}}
    hasChanges={false}
    onChange={[Function]}
    onComplete={[Function]}
    rules={Object {}}
    show={false}
    stageId={null}
    updateDraft={[Function]}
    updateStage={[Function]}
    variableRegistry={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "buttons": Array [],
      "children": <SkipLogicEditor
        onChange={[Function]}
        rules={Object {}}
        variableRegistry={Object {}}
      />,
      "onCancel": [Function],
      "show": false,
      "type": "logic",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "onChange": [Function],
        "rules": Object {},
        "variableRegistry": Object {},
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "buttons": Array [],
        "children": <SkipLogicEditor
          onChange={[Function]}
          rules={Object {}}
          variableRegistry={Object {}}
        />,
        "onCancel": [Function],
        "show": false,
        "type": "logic",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onChange": [Function],
          "rules": Object {},
          "variableRegistry": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
